clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -2918,24 +2919,60 @@ static bool checkWorkGroupSizeValues(Sema &S, Decl *D, const ParsedAttr &Attr,
   };
 
   if (Attr.getKind() == ParsedAttr::AT_SYCLIntelMaxGlobalWorkDim) {
-    if (const auto *A = D->getAttr<SYCLIntelMaxWorkGroupSizeAttr>())
-      Result &= checkZeroDim(A, A->getXDim(), A->getYDim(), A->getZDim());
-    if (const auto *A = D->getAttr<ReqdWorkGroupSizeAttr>())
-      Result &= checkZeroDim(A, A->getXDim(), A->getYDim(), A->getZDim());
+    if (const auto *A = D->getAttr<SYCLIntelMaxWorkGroupSizeAttr>()) {
```

Comment:
I'm not sure if I am missing something but it looks like there is a some code duplication here. Can we combine the if-else block here by using hasAttr<xyz>() || hasAttr<abc>(). 

I guess however this means you will have a getAttr() after hasAttr() which @AaronBallman mentioned is a bad coding practice. So I'm not sure. Tagging @AaronBallman

Reply:
If we could use C++20 features, this would be easier to remove the duplication from because we could use a templated lambda. As it stands, I think the code is reasonable enough for the moment (we'd need the attributes to have a common base class otherwise, and I'm not certain if that's reasonable or not).

Reply:
Yes, that 's what i thought to have bad coding style if we have getAttr after hasAttr(). I did not make any change here. 

@AaronBallman, any thought here since we decided to remove getIntExprValue() in our previous discussion (that removed some code duplication here).

Reply:
Heh, I think we're all reviewing the same code at roughly the same time, so some of our comments come across in a fun order. :-D

There are a few approaches we can consider, but y'all know the domain better than I do and so some of these suggestions may be a bad idea. If these attributes are doing the same thing but are just spelled differently, we could use a single semantic attribute in Attd.td with different spellings. Then anywhere we do `D->getAttr<>()` there'd only be the one semantic attribute to query (and we could add an accessor to it if we need to know which spelling the attribute was parsed with). If the two attributes are not doing the same thing but instead are doing highly related things, then perhaps a common attribute base class that both of the semantic attributes inherit from would be a good approach. But if the attributes are doing drastically different things, then I think this code is sufficient as-is.

Reply:
The attributes are doing different things here:

1. max_global_work_dim() attribute take only single argument
     reqd_work_group_size, max_work_group_size take triple attribute argument values.
2. Values of reqd_work_group_size arguments shall be equal or less than values
    coming from max_work_group_size.
3. In case the value of 'max_global_work_dim' attribute equals to 0 we shall
ensure that if max_work_group_size and reqd_work_group_size attributes exist,
 they hold equal values (1, 1, 1).
4. ReqdWorkGroupSizeAttr gives conflicts errors in case parsed attribute value >= Decl attribute.


2. if (const auto *A = D->getAttr<ReqdWorkGroupSizeAttr>()) {
    if (!(WGSize[0] >= A->getXDim() && WGSize[1] >= A->getYDim() &&
          WGSize[2] >= A->getZDim())) {

Reply:
Am I understanding properly that `reqd_work_group_size` and `max_work_group_size` are both specifying similar properties? Namely that one is a request and the other is the max request that can be fulfilled?

If so, would it make sense for there to be a single semantic attribute named `WorkGroupSizeAttr` that looks something like this:
```
def WorkGroupSize : InheritableAttr {
  let Spellings = [CXX11<"intelfpga","max_work_group_size">,
                   CXX11<"intel","max_work_group_size">,
                   GNU<"reqd_work_group_size">,
                   CXX11<"intel","reqd_work_group_size">,
                   CXX11<"cl","reqd_work_group_size">];
  let Args = [ExprArgument<"XDim">,
              ExprArgument<"YDim">,
              ExprArgument<"ZDim">];
  let Subjects = SubjectList<[Function], ErrorDiag>;
  let Accessors = [Accessor<"isMax",
                            [CXX11<"intelfpga","max_work_group_size">,
                             CXX11<"intel","max_work_group_size">]>,
                   Accessor<"isRequest",
                            [GNU<"reqd_work_group_size">,
                             CXX11<"intel","reqd_work_group_size">,
                             CXX11<"cl","reqd_work_group_size">]>];
}
```
Then all of the code using the semantic attributes would use `WorkGroupSizeAttr` instead, and if they need to care about max vs reqd, they can call `isMax()` or `isRequest()` as needed.

(Note, I'm speculating here. This idea may or may not make sense.)

Reply:
NOTE: There are few differences between them:
1. "max_work_group_size" is similar to reqd_work_group_size, but allows work groups that are smaller or equal to the specified sizes.
2. Default arguments in ReqWorkGroupSize can be used only with
intel::reqd_work_group_size spelling.
3. ReqdWorkGroupSizeAttr gives conflicts errors in case parsed attribute value >= Decl attribute but not "max_work_group_size".

Reply:
Another idea that came to me last night while I should have been sleeping was that we could add an additional member to provide an array interface to the attribute arguments. This would at least let you do things in a loop where you're currently repeating the same code once for each dimension. Something like:
```
let AdditionalMembers = [{
  ArrayRef<const Expr *> dimensions() const {
    return {getXDim(), getYDim(), getZDim()};
  }
  ArrayRef<Expr *> dimensions() {
    return {getXDim(), getYDim(), getZDim()};
  }
}];
```
and this would allow you to do something like:
```
for (auto Item : zip(OldDeclAttr->dimensions(), NewDeclAttr->dimensions()) {
  Expr *Old = std::get<0>(Item), *New = std::get<1>(Item);
  ...
}
```
or other such things. I've not tested the code out, but it seems like a plausible improvement that doesn't muck about with the type hierarchy.

Reply:
Done